TypeCobol/Compiler/Scopes/Scope.cs

```diff
@@ -0,0 +1,378 @@
+            /// </summary>
+            public override int Count => _symbols.Count;
+
+            public override T this[int i] => _symbols[i];
+
+            /// <summary>
+            /// Remove the given symbol in this multy entries
+            /// </summary>
+            /// <param name="sym"></param>
+            internal void Remove(T sym)
+            {
+                //After removing a multi must at least contains one element.
+                System.Diagnostics.Contracts.Contract.Ensures(Count >= 1);
+                _symbols.Remove(sym);                
+                //base.Symbol is always the first symbol of the list.
+                Symbol = _symbols[0];
+            }
+        }
+
+
+        /// <summary>
+        /// 
+        /// The Owner of this scope
+        /// </summary>
+        public Symbol Owner
+        {
+            get;
+            set;
+        }
+        /// <summary>
+        /// All Symbols in this scope, each symbol is associated an integer number which is
+        /// its index in the _OrderedSymbols list.
+        /// </summary>
+        Dictionary<String, Entry > _symbols = null;
+        /// <summary>
+        /// All symbol in enter order.
+        /// </summary>
+        List<T> _orderedSymbols = null;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="owner">Scope Owner</param>
+        public Scope(Symbol owner)
+        {
+            this.Owner = owner;
+        }
+
+        /// <summary>
+        /// Enter a Symbol in this scope, if a previous symbol with the same name exits, then a MultySymbols entry
+        /// will be used to store the nes symbol.
+        /// </summary>
+        /// <param name="sym">The Symbol to enter</param>
+        /// <returns>The entry of the symbol</returns>
+        public Entry Enter(T sym)
+        {
+            System.Diagnostics.Debug.Assert(sym != null);
+            System.Diagnostics.Debug.Assert(sym.Name != null);
+            if (_symbols == null)
+            {
+                _symbols = new Dictionary<String, Entry >(StringComparer.OrdinalIgnoreCase);
+                _orderedSymbols = new List<T>();
+            }
+            Entry entry = null;
+            String name = sym.Name;
+            if (!_symbols.ContainsKey(name))
+            {
+                _symbols[name] = entry = new Entry(sym);
+            }
+            else
+            {
+                //Check if the symbol itself exits                
+                entry = _symbols[name];
+                if (entry.Exits(sym))
+                    return entry;
+                MultiSymbols multi = null;
+                if (entry.Count == 1)
+                {//Create a multi symbols entry
+                    multi = new MultiSymbols(entry.Symbol);                    
+                    _symbols[name] = multi;
+                }
+                else
+                {
+                    multi = _symbols[name] as MultiSymbols;
```

Comment:
Better use `multi = entry as MultiSymbols;` to avoid another dictionary lookup.

Reply:
ok will be replaced by multi = (MultiSymbols)entry;